python/hsfs/core/arrow_flight_client.py

```diff
@@ -0,0 +1,335 @@
+    def read_path(self, path):
+        descriptor = pyarrow.flight.FlightDescriptor.for_path(path)
+        return self._get_dataset(descriptor)
+
+    @_handle_afs_exception
+    def create_training_dataset(self, feature_view, tds_version=1):
+        if not self._is_enabled:
+            raise FeatureStoreException("Arrow Flight Service is not enabled.")
+        training_dataset_metadata = self._training_dataset_metadata_from_feature_view(
+            feature_view, tds_version
+        )
+        training_dataset_encoded = json.dumps(training_dataset_metadata).encode("ascii")
+        buf = pyarrow.py_buffer(training_dataset_encoded)
+        action = pyarrow.flight.Action("create-training-dataset", buf)
+        for result in self._connection.do_action(action):
+            return result.body.to_pybytes()
+
+    def _path_from_feature_view(self, feature_view, tds_version):
+        training_dataset_metadata = self._training_dataset_metadata_from_feature_view(
+            feature_view, tds_version
+        )
+        path = (
+            f"{training_dataset_metadata['featurestore_name']}_Training_Datasets/"
+            f"{training_dataset_metadata['name']}_{training_dataset_metadata['version']}"
+            f"_{training_dataset_metadata['tds_version']}/"
+            f"{training_dataset_metadata['name']}_{training_dataset_metadata['version']}/"
+            f"{training_dataset_metadata['name']}_{training_dataset_metadata['version']}.parquet"
+        )
+        full_path = f"/Projects/{training_dataset_metadata['featurestore_name']}/{path}"
+        return full_path
+
+    def _training_dataset_metadata_from_feature_view(self, feature_view, tds_version):
+        training_dataset_metadata = {}
+        training_dataset_metadata["name"] = feature_view.name
+        training_dataset_metadata["version"] = f"{feature_view.version}"
+        training_dataset_metadata["tds_version"] = f"{tds_version}"
+        training_dataset_metadata["query"] = self._get_query_object(
+            feature_view.query, feature_view.query.to_string()
+        )
+        training_dataset_metadata[
+            "featurestore_name"
+        ] = feature_view.query._left_feature_group.feature_store_name.replace(
+            "_featurestore", ""
+        )
+
+        return training_dataset_metadata
+
+    def _get_query_object(self, query, query_str):
+        query_string = query_str.replace(
+            f"`{query._left_feature_group.feature_store_name}`.`",
+            f"`{query._left_feature_group.feature_store_name.replace('_featurestore','')}.",
+        ).replace("`", '"')
+
+        (
+            featuregroups,
+            features,
+            filters,
+        ) = self._collect_featuregroups_features_and_filters(query)
+
+        query = {
+            "query_string": query_string,
+            "featuregroups": featuregroups,
+            "features": features,
+            "filters": filters,
+        }
+        return query
```

Comment:
```suggestion
        training_dataset_metadata["query"] = self._get_query_object(
            feature_view.query
        )
        training_dataset_metadata[
            "featurestore_name"
        ] = feature_view.query._left_feature_group.feature_store_name.replace(
            "_featurestore", ""
        )

        return training_dataset_metadata

    def _get_query_object(self, query):
        query_string = query.to_string().replace(
            f"`{query._left_feature_group.feature_store_name}`.`",
            f"`{query._left_feature_group.feature_store_name.replace('_featurestore','')}.",
        ).replace("`", '"')

        (
            featuregroups,
            features,
            filters,
        ) = self._collect_featuregroups_features_and_filters(query)

        query = {
            "query_string": query_string,
            "featuregroups": featuregroups,
            "features": features,
            "filters": filters,
        }
        return query
```

Reply:
I know it would be nice, but `read_query` needs to call `_get_query_object` as well, and it will get the query and the query string as separate objects.

Reply:
I will try to do some more refactoring.